plugin.go

```diff
@@ -0,0 +1,391 @@
+}
+
+// PluginUser .This is a TBD Comments.
+//
+// See https://goo.gl/kaOHGw for more details.
+type PluginUser struct {
+	UID int32 `json:"UID,omitempty" yaml:"UID,omitempty" toml:"UID,omitempty"`
+	GID int32 `json:"GID,omitempty" yaml:"GID,omitempty" toml:"GID,omitempty"`
+}
+
+// PluginConfig .This is a TBD Comments.
+//
+// See https://goo.gl/kaOHGw for more details.
+type PluginConfig struct {
+	Description     string `json:"Description,omitempty" yaml:"Description,omitempty" toml:"Description,omitempty"`
+	Documentation   string
+	Interface       PluginInterface `json:"Interface,omitempty" yaml:"Interface,omitempty" toml:"Interface,omitempty"`
+	Entrypoint      []string        `json:"Entrypoint,omitempty" yaml:"Entrypoint,omitempty" toml:"Entrypoint,omitempty"`
+	WorkDir         string          `json:"WorkDir,omitempty" yaml:"WorkDir,omitempty" toml:"WorkDir,omitempty"`
+	User            PluginUser      `json:"User,omitempty" yaml:"User,omitempty" toml:"User,omitempty"`
+	Network         PluginNetwork   `json:"Network,omitempty" yaml:"Network,omitempty" toml:"Network,omitempty"`
+	Linux           PluginLinux     `json:"Linux,omitempty" yaml:"Linux,omitempty" toml:"Linux,omitempty"`
+	PropagatedMount string          `json:"PropagatedMount,omitempty" yaml:"PropagatedMount,omitempty" toml:"PropagatedMount,omitempty"`
+	Mounts          []Mount         `json:"Mounts,omitempty" yaml:"Mounts,omitempty" toml:"Mounts,omitempty"`
+	Env             []PluginEnv     `json:"Env,omitempty" yaml:"Env,omitempty" toml:"Env,omitempty"`
+	Args            PluginArgs      `json:"Args,omitempty" yaml:"Args,omitempty" toml:"Args,omitempty"`
+}
+
+// PluginDetail .This is a TBD Comments.
+//
+// See https://goo.gl/kaOHGw for more details.
+type PluginDetail struct {
+	ID       string        `json:"Id,omitempty" yaml:"Id,omitempty" toml:"Id,omitempty"`
+	Name     string        `json:"Name,omitempty" yaml:"Name,omitempty" toml:"Name,omitempty"`
+	Tag      string        `json:"Tag,omitempty" yaml:"Tag,omitempty" toml:"Tag,omitempty"`
+	Active   bool          `json:"Active,omitempty" yaml:"Active,omitempty" toml:"Active,omitempty"`
+	Settings PluginSetting `json:"Settings,omitempty" yaml:"Settings,omitempty" toml:"Settings,omitempty"`
+	Config   PluginConfig  `json:"Config,omitempty" yaml:"Config,omitempty" toml:"Config,omitempty"`
+}
+
+// ListPlugins .This is a TBD Comments.
+//
+// See https://goo.gl/kaOHGw for more details.
+func (c *Client) ListPlugins() ([]PluginDetail, error) {
+	resp, err := c.do("GET", "/plugins", doOptions{})
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	pluginDetails := make([]PluginDetail, 0)
+	if err := json.NewDecoder(resp.Body).Decode(&pluginDetails); err != nil {
+		return nil, err
+	}
+	return pluginDetails, nil
+}
+
+// GetPluginPrivileges .This is a TBD Comments.
+//
+// See https://goo.gl/kaOHGw for more details.
+func (c *Client) GetPluginPrivileges(name string) ([]PluginPrivilege, error) {
```

Comment:
Can you  make sure that this and all other methods get a ``Context``? I tagged them with "Missing ``Context``." to make it easier for you to track them as you work on the changes. The idea is to always support canceling in flight requests if the context gets canceled.

Please use a struct (``GetPluginPrivilegesOptions``), as it makes it easier to add new options as Docker evolves their API.